keps/sig-network/2136-np-version-status/README.md

```diff
@@ -0,0 +1,789 @@
+the policy is _completely_ unsupported. Unless there is also a `{ type:
+"Enforcing", status: "False" }` condition, it is possible that the
+policy may be partially-enforced.
+
+  - If the policy's `spec.minVersion` is a version that the plugin does
+    not recognize, then the plugin should set a status condition with `{
+    type: "Supported", status: "False", reason: "Version" }` and an
+    appropriate `message` (and likewise it should set an appropriate
+    `"Enforcing"` condition indicating that it is not enforcing the
+    policy).
+
+  - If the policy has a recognized `minVersion` (or no `minVersion`),
+    but the plugin can see that it uses features which the plugin does
+    not support, then the plugin should set a status condition with `{
+    type: "Supported", status: "False", reason: "Unimplemented" }` and
+    an appropriate `message`. It is up to the plugin's discretion
+    whether it should also set a `"False"` `"Enforcing"` status like
+    above, or if it instead chooses to implement as much of the
+    NetworkPolicy as it can.
+
+  - If a policy has a recognized `minVersion` and uses only supported
+    features, the plugin _may_ set a status condition with `{ type:
+    "Supported", status: "True" }`. However, this can also be assumed as
+    the default value if there is no `"Supported"` condition.
+
+  - If a policy has no `minVersion`, but appears to use only supported
+    features, the plugin _may_ set a status condition with `{ type:
+    "Supported", status: "Unknown", reason: "NoMinVersion" }`. (In the
+    absence of `minVersion`, it is impossible for the plugin to know for
+    sure that the policy does not use some new feature that the plugin
+    is unaware of.)
+
+Note the distinction between the first two cases; if the `minVersion` is
+higher than the highest version the plugin knows about, then the plugin
+has no way of telling whether it can implement the policy or not, and no
+way of knowing for sure if it's correctly interpreting even the parts
+that it can see. (For example, if a network plugin that only implemented
+version "1.9" received a policy with a combined `podSelector` and
+`namespaceSelector`, it might misinterpret it by reading the
+`namespaceSelector` but not the `podSelector`, and thereby allowing
+access to/from more pods than was intended.) Thus, it is best for it to
+not even try.
+
+On the other hand, if the plugin knows about a feature but doesn't
+implement it, they may still be able to implement the other parts of the
+policy safely. (For example, a plugin that does not implement egress
+policies at all might choose to still implement the ingress side of a
+mixed ingress/egress policy.)
+
+#### The "Enforcing" Condition
+
+The `"Enforcing"` condition indicates whether the plugin is currently
+enforcing the policy described by a NetworkPolicy.
+
+- If the plugin is intentionally not enforcing the policy (eg, because
```

Comment:
this is great for the loopback policy condition, which may or may not be enforced based
on wether kube-proxy is being used in one manner or other.  cc @mattfenwick 

it would also allow us possibly to actually define wether loopback policy is supported or not w/o breaking anything in the validation

Reply:
(Clarifying: the issue here is that it's unspecified whether an isolated pod is implicitly allowed to connect to itself, or if it can only connect to itself when there is a policy explicitly allowing that. With most plugins, self-connection is implicitly allowed, because NPs are enforced outside of the pod's namespace, so a self-connection never hits the NP rules. But with some plugins, like Cilium, isolated pods are _not_ allowed to connect to themselves unless there is a policy allowing it.)

I'm not sure anything currently in this proposal is a good way of explaining this though. For one, if we did expect plugins to indicate whether loopback connections were or were not being blocked, then they would have to set that condition on basically every NetworkPolicy object, since it is not a property of any specific NetworkPolicy. (eg, if you have a policy that says "allow from X to Y" then that also either does or does not imply "block from Y to Y" depending on NP implementation.)

This feels more like a global implementation detail than a per-policy status. I had thought about whether we should have some global status in addition to the per-policy statuses. eg, then you could know for sure if this was a plugin that implements NetworkPolicyStatus or not... Also, maybe it could set a `supportedVersion` field so you could know ahead of time what things were supported...

Reply:
ah, yeah, so i guess, publishing " global provider metadata " as part of this policy status field is an impedence mismatch ....

... selfishly,  i would like that global metadata (or just hacking the metadata into this stuff asis) from that "user story #1" perspective just bc then we could validate loopback in netpol matrices .   

But if you decide that doesnt fit in here, thats fine to.  just being selfish :)

Reply:
> But if you decide that doesnt fit in here

well, I was saying I don't think it fits into the API that's _currently_ in the KEP, but there are definitely arguments for adding something to cover that as well

Reply:
The loopback loophole seems orthogonal - we should define what we expect it to do and then push that to implementations, with a CVE if needed